### PR TITLE
CleanUriSegment optimization.

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR includes optimizations to the `CleanUriSegment` method by introducing a no allocating span algorithm for `netcoreapp`, a `StringBuilderCache` based on the one in the BCL and prioritize chars over string in the Append method. Also includes missing unit tests.

#### Some benchmarks:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.630 (2004/?/20H1)
Intel Core i7-1068NG7 CPU 2.30GHz, 1 CPU, 2 logical and 2 physical cores
.NET Core SDK=5.0.100
  [Host]     : .NET Core 3.1.9 (CoreCLR 4.700.20.47201, CoreFX 4.700.20.47203), X64 RyuJIT
  Job-PZAJGD : .NET Framework 4.8 (4.8.4250.0), X64 RyuJIT
  Job-CNRSOZ : .NET Core 3.1.9 (CoreCLR 4.700.20.47201, CoreFX 4.700.20.47203), X64 RyuJIT


```
|             Method |     Toolchain |         absolutePath |       Mean |      Error |     StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |-------------- |--------------------- |-----------:|-----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|    CleanUriSegment |        net472 |                    / |   4.040 ns |  0.0702 ns |  0.0622 ns |  1.00 |    0.00 |      - |     - |     - |         - |
|    CleanUriSegment | netcoreapp3.1 |                    / |   4.010 ns |  0.0392 ns |  0.0367 ns |  0.99 |    0.02 |      - |     - |     - |         - |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 |                    / |   4.144 ns |  0.0295 ns |  0.0246 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| CleanUriSegmentNew | netcoreapp3.1 |                    / |   4.026 ns |  0.1097 ns |  0.1347 ns |  0.95 |    0.03 |      - |     - |     - |         - |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 |  /Dat(...)s.cs [107] | 699.948 ns | 12.0800 ns | 11.2997 ns |  1.00 |    0.00 | 0.2289 |     - |     - |     963 B |
|    CleanUriSegment | netcoreapp3.1 |  /Dat(...)s.cs [107] | 358.188 ns |  6.8910 ns |  6.7679 ns |  0.51 |    0.01 | 0.2198 |     - |     - |     920 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 |  /Dat(...)s.cs [107] | 704.657 ns | 14.0551 ns | 16.7316 ns |  1.00 |    0.00 | 0.1602 |     - |     - |     674 B |
| CleanUriSegmentNew | netcoreapp3.1 |  /Dat(...)s.cs [107] | 222.460 ns |  3.4643 ns |  3.2405 ns |  0.32 |    0.01 | 0.0572 |     - |     - |     240 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 | /cont(...)66d4d [55] | 647.279 ns | 11.3619 ns | 10.6279 ns |  1.00 |    0.00 | 0.1068 |     - |     - |     449 B |
|    CleanUriSegment | netcoreapp3.1 | /cont(...)66d4d [55] | 247.928 ns |  4.8938 ns |  8.1764 ns |  0.39 |    0.01 | 0.1030 |     - |     - |     432 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 | /cont(...)66d4d [55] | 648.606 ns | 12.8090 ns | 11.9816 ns |  1.00 |    0.00 | 0.0629 |     - |     - |     265 B |
| CleanUriSegmentNew | netcoreapp3.1 | /cont(...)66d4d [55] | 190.178 ns |  2.9059 ns |  2.7182 ns |  0.29 |    0.01 | 0.0153 |     - |     - |      64 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 | /cont(...)66d4d [55] | 449.936 ns |  5.9746 ns |  5.2963 ns |  1.00 |    0.00 | 0.1221 |     - |     - |     514 B |
|    CleanUriSegment | netcoreapp3.1 | /cont(...)66d4d [55] | 197.503 ns |  3.5414 ns |  3.3126 ns |  0.44 |    0.01 | 0.1204 |     - |     - |     504 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 | /cont(...)66d4d [55] | 451.972 ns |  8.9948 ns |  8.4137 ns |  1.00 |    0.00 | 0.0782 |     - |     - |     329 B |
| CleanUriSegmentNew | netcoreapp3.1 | /cont(...)66d4d [55] | 136.092 ns |  2.1910 ns |  2.0495 ns |  0.30 |    0.01 | 0.0324 |     - |     - |     136 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 | /cont(...)66d4d [55] | 451.443 ns |  6.6462 ns |  6.2168 ns |  1.00 |    0.00 | 0.1221 |     - |     - |     514 B |
|    CleanUriSegment | netcoreapp3.1 | /cont(...)66d4d [55] | 198.143 ns |  2.3514 ns |  1.9636 ns |  0.44 |    0.01 | 0.1204 |     - |     - |     504 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 | /cont(...)66d4d [55] | 446.901 ns |  7.1954 ns |  6.7306 ns |  1.00 |    0.00 | 0.0782 |     - |     - |     329 B |
| CleanUriSegmentNew | netcoreapp3.1 | /cont(...)66d4d [55] | 133.722 ns |  1.9480 ns |  1.8221 ns |  0.30 |    0.01 | 0.0324 |     - |     - |     136 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 | /cont(...)66d4d [55] | 445.394 ns |  7.2481 ns |  6.7799 ns |  1.00 |    0.00 | 0.1221 |     - |     - |     514 B |
|    CleanUriSegment | netcoreapp3.1 | /cont(...)66d4d [55] | 203.372 ns |  3.2887 ns |  2.9154 ns |  0.46 |    0.01 | 0.1204 |     - |     - |     504 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 | /cont(...)66d4d [55] | 447.919 ns |  6.8444 ns |  7.6075 ns |  1.00 |    0.00 | 0.0782 |     - |     - |     329 B |
| CleanUriSegmentNew | netcoreapp3.1 | /cont(...)66d4d [55] | 135.899 ns |  2.6486 ns |  3.3496 ns |  0.30 |    0.01 | 0.0324 |     - |     - |     136 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
|    CleanUriSegment |        net472 | /cont(...)d334e [51] | 749.808 ns | 14.3617 ns | 13.4340 ns |  1.00 |    0.00 | 0.1431 |     - |     - |     602 B |
|    CleanUriSegment | netcoreapp3.1 | /cont(...)d334e [51] | 236.155 ns |  3.6571 ns |  3.7556 ns |  0.31 |    0.01 | 0.0992 |     - |     - |     416 B |
|                    |               |                      |            |            |            |       |         |        |       |       |           |
| CleanUriSegmentNew |        net472 | /cont(...)d334e [51] | 743.738 ns | 13.8776 ns | 13.6297 ns |  1.00 |    0.00 | 0.1011 |     - |     - |     425 B |
| CleanUriSegmentNew | netcoreapp3.1 | /cont(...)d334e [51] | 186.414 ns |  3.3248 ns |  3.1100 ns |  0.25 |    0.01 | 0.0153 |     - |     - |      64 B |

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace/Util/UriHelpers.cs [hunk 1]
```diff
@@ -42,8 +40,48 @@ public static string CleanUriSegment(string absolutePath)
...
+            ReadOnlySpan<char> absPath = absolutePath.AsSpan();
+
             // Sanitized url will be at worse as long as the original
-            var sb = new StringBuilder(absolutePath.Length);
+            var sb = StringBuilderCache.Acquire(absPath.Length);
+
+            int previousIndex = 0;
+            int index;
+
+            do
```
Comment (kevingosse) [inline src/Datadog.Trace/Util/UriHelpers.cs:52]: This makes me realize that the loop in the .net framework version should have been a do/while as well

**Review by kevingosse — APPROVED** [review]
LGTM

**Review by zacharycmontoya — APPROVED** [review]

File: src/Datadog.Trace/Util/UriHelpers.cs (hunk 1 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Util/UriHelpers.cs:52]: Doesn't look like we made this change, @tonyredondo.

File: test/Datadog.Trace.Tests/Util/CleanUriSegmentTests.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    public class CleanUriSegmentTests
+    {
+        [Theory]
+        [InlineData("/", "/")]
+        [InlineData("/controller/action/b37855d4bae34bd3b3357fc554ad334e", "/controller/action/?")]
+        [InlineData("/controller/action/14bb2eed-34f0-4aa2-b2c3-09c0e2166d4d", "/controller/action/?")]
+        [InlineData("/controller/action/14bb2eed-34f0X4aa2-b2c3-09c0e2166d4d", "/controller/action/14bb2eed-34f0X4aa2-b2c3-09c0e2166d4d")]
+        [InlineData("/controller/action/14bb2eed-34f0-4aa2Xb2c3-09c0e2166d4d", "/controller/action/14bb2eed-34f0-4aa2Xb2c3-09c0e2166d4d")]
+        [InlineData("/controller/action/14bb2eed-34f0A4aa2Bb2c3C09c0e2166d4d", "/controller/action/14bb2eed-34f0A4aa2Bb2c3C09c0e2166d4d")]
+        [InlineData("/DataDog/dd-trace-dotnet/blob/e2d83dec7d6862d4181937776ddaf72819e291ce/src/Datadog.Trace/Util/UriHelpers.cs", "/DataDog/dd-trace-dotnet/blob/e2d83dec7d6862d4181937776ddaf72819e291ce/src/Datadog.Trace/Util/UriHelpers.cs")]
```
Comment (lucaspimentel) [inline test/Datadog.Trace.Tests/Util/CleanUriSegmentTests.cs:14]: This is out of scope of this PR, but this test data makes me think we should replace these other cases as well, even though they don't specifically parse as GUIDs or integers. Maybe we should expand our logic to any hexadecimal number as well (which any number/position of dashes).

For example, the commit hash in
```
/DataDog/dd-trace-dotnet/blob/e2d83dec7d6862d4181937776ddaf72819e291ce/src/Datadog.Trace/Util/UriHelpers.cs
```
should be removed to reduce resource name cardinality:
```
/DataDog/dd-trace-dotnet/blob/?/src/Datadog.Trace/Util/UriHelpers.cs
```

**Review by lucaspimentel — COMMENTED** [review]
Late to the party 😅

File: src/Datadog.Trace/Util/UriHelpers.cs (hunk 1 above)
Comment (tonyredondo) [reply src/Datadog.Trace/Util/UriHelpers.cs:52]: Oh, you're right, I'll create a new PR with that change.

File: test/Datadog.Trace.Tests/Util/CleanUriSegmentTests.cs (hunk 2 above)
Comment (lucaspimentel) [reply test/Datadog.Trace.Tests/Util/CleanUriSegmentTests.cs:14]: Another thought is to add a user setting where they can specify a regex to match custom urls. For example something like `/country/spain/edit` may be `/country/?/edit`, but we would never catch that without user input or a route template.

Anyway, just something to think about.

Comment (tonyredondo) [reply test/Datadog.Trace.Tests/Util/CleanUriSegmentTests.cs:14]: I agree with both ideas. A little concern in performance impact supporting regex values but yeah, I think is useful a feature like that.

**Pull request merged**